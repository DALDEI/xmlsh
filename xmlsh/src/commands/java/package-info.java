/**
 * 
 */
/**
 * @author David A. Lee
 */
package java;