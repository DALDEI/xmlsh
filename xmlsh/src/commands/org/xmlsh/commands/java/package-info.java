/**
 * 
 */
/**
 * @author David A. Lee
 */
package org.xmlsh.commands.java;