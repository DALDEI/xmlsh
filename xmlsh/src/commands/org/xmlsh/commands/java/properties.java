/**
 * $Id: $
 * $Date: $
 *
 */

package org.xmlsh.commands.java;

import org.xmlsh.core.BuiltinFunctionCommand;
import org.xmlsh.core.CoreException;
import org.xmlsh.core.InvalidArgumentException;
import org.xmlsh.core.Options;
import org.xmlsh.core.XValue;
import org.xmlsh.sh.shell.SerializeOpts;
import org.xmlsh.sh.shell.Shell;
import org.xmlsh.types.TypeFamily;
import org.xmlsh.util.StringPair;

import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.util.List;
import java.util.Properties;

public class properties extends BuiltinFunctionCommand
{

	private SerializeOpts sopts;

	public properties()
	{
		super("properties");
	}

	@Override
	public XValue run(Shell shell, List<XValue> args) throws Exception
	{

		Options opts = new Options("t=text,x=xml,f=file,m=map,v=variables,p=pairs");
		opts.parse(args);

		boolean bText = opts.hasOpt("t");   // parse in text format
		boolean bXml = opts.hasOpt("x");    // parse in XML format
		boolean bMap = opts.hasOpt("m");    // parse in Map format
		boolean bVars = opts.hasOpt("v");   // parse pairs of vars
		boolean bPairs = opts.hasOpt("p");  // parse a=b pairs
		boolean bFile  = opts.hasOpt("file") ;  // input from file 

		args = opts.getRemainingArgs();

		Properties props = new Properties();

		if( args.isEmpty() )
			return newValue(props);



		sopts = shell.getSerializeOpts();
		if( bText || bXml ) {

			try( InputStream is = getInputStream( shell , bFile , args , sopts )){
				if( bText )
					props.load(is);
				else
					props.loadFromXML(is);

			}
		}
		else	

			if( bVars ) {
				for( int i = 0 ; i < args.size() - 1 ; i+=2 ) {
					props.put( args.get(i).toString() , args.get(i+1).toString());
				}
			}
			else
				if( bPairs ) {
					for (XValue add : args ){
						StringPair pair = new StringPair( add.toString() , '=');
						props.setProperty( pair.getLeft(), pair.getRight() );
					}
				}

		return newValue(props);

	}

	private InputStream getInputStream(Shell shell , boolean bFile , List<XValue> args, SerializeOpts sopts) throws InvalidArgumentException, IOException, CoreException
	{
		if( bFile   )
			return shell.getEnv().getInput( args.get(0) ).asInputStream(sopts);
		else {
			XValue a = args.get(0);
			return new ByteArrayInputStream( a.toByteArray(sopts));

		}
	}

	public XValue newValue(Properties props)
	{
		return new XValue( TypeFamily.JAVA , props );
	}


}



/*
 * Copyright (C) 2008-2012 David A. Lee.
 * 
 * The contents of this file are subject to the "Simplified BSD License" (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy of the
 * License at http://www.opensource.org/licenses/bsd-license.php 

 * Software distributed under the License is distributed on an "AS IS" basis,
 * WITHOUT WARRANTY OF ANY KIND, either express or implied.
 * See the License for the specific language governing rights and limitations under the License.
 *
 * The Original Code is: all this file.
 *
 * The Initial Developer of the Original Code is David A. Lee
 *
 * Portions created by (your name) are Copyright (C) (your legal entity). All Rights Reserved.
 *
 * Contributor(s): David A. Lee
 * 
 */