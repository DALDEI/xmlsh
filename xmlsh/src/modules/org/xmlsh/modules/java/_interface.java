/**
 * $Id: $
 * $Date: $
 *
 */

package org.xmlsh.modules.java;

import java.util.Collections;
import java.util.List;

import net.bytebuddy.ByteBuddy;
import net.bytebuddy.dynamic.ClassLoadingStrategy;
import net.bytebuddy.dynamic.DynamicType;
import net.bytebuddy.instrumentation.FieldAccessor;
import net.bytebuddy.modifier.Visibility;
import static net.bytebuddy.matcher.ElementMatchers.*;
import static net.bytebuddy.utility.ByteBuddyCommons.*;
import net.bytebuddy.asm.ClassVisitorWrapper;
import net.bytebuddy.dynamic.ClassFileLocator;
import net.bytebuddy.dynamic.DynamicType;
import net.bytebuddy.dynamic.scaffold.BridgeMethodResolver;
import net.bytebuddy.dynamic.scaffold.FieldRegistry;
import net.bytebuddy.dynamic.scaffold.MethodRegistry;
import net.bytebuddy.dynamic.scaffold.inline.InlineDynamicTypeBuilder;
import net.bytebuddy.dynamic.scaffold.inline.MethodRebaseResolver;
import net.bytebuddy.dynamic.scaffold.subclass.ConstructorStrategy;
import net.bytebuddy.dynamic.scaffold.subclass.SubclassDynamicTypeBuilder;
import net.bytebuddy.instrumentation.Instrumentation;
import net.bytebuddy.instrumentation.ModifierContributor;
import net.bytebuddy.instrumentation.attribute.FieldAttributeAppender;
import net.bytebuddy.instrumentation.attribute.MethodAttributeAppender;
import net.bytebuddy.instrumentation.attribute.TypeAttributeAppender;
import net.bytebuddy.instrumentation.method.MethodDescription;
import net.bytebuddy.instrumentation.method.MethodLookupEngine;
import net.bytebuddy.instrumentation.type.TypeDescription;
import net.bytebuddy.instrumentation.type.TypeList;
import net.bytebuddy.matcher.ElementMatcher;
import net.bytebuddy.modifier.TypeManifestation;

import org.xmlsh.annotations.Function;
import org.xmlsh.core.AbstractBuiltinFunction;
import org.xmlsh.core.XValue;
import org.xmlsh.sh.shell.Shell;
import org.xmlsh.util.JavaUtils;

@Function(name="interface",names={"new-interface"} )
public class _interface extends AbstractBuiltinFunction {
    @SuppressWarnings({ "unchecked" })
    @Override
	public XValue run(Shell shell, List<XValue> args) throws Exception {

	    String name = args.remove(0).toString();
	    
	    DynamicType.Builder<?> b = new ByteBuddy().
	            makeInterface().name(name);

	    for( XValue a : args ){
	        String f = a.toString();
            b=b.defineMethod( JavaUtils.toSetterName(f)  ,Void.TYPE, ( List<Class<?>> )(List)  Collections.
                    singletonList(String.class) , Visibility.PUBLIC).withoutCode().
                defineMethod( JavaUtils.toGetterName(f) , String.class ,( List<Class<?>> ) (List) Collections.emptyList() , Visibility.PUBLIC ).withoutCode();
	    }

	    Class cls = b.make()
	                .load(getClass().getClassLoader(), ClassLoadingStrategy.Default.WRAPPER)
	                .getLoaded();

	   		return XValue.newXValue( cls  );


	}

}



//
//
//Copyright (C) 2008-2014 David A. Lee.
//
//The contents of this file are subject to the "Simplified BSD License" (the "License");
//you may not use this file except in compliance with the License. You may obtain a copy of the
//License at http://www.opensource.org/licenses/bsd-license.php 
//
//Software distributed under the License is distributed on an "AS IS" basis,
//WITHOUT WARRANTY OF ANY KIND, either express or implied.
//See the License for the specific language governing rights and limitations under the License.
//
//The Original Code is: all this file.
//
//The Initial Developer of the Original Code is David A. Lee
//
//Portions created by (your name) are Copyright (C) (your legal entity). All Rights Reserved.
//
//Contributor(s): none.
//
