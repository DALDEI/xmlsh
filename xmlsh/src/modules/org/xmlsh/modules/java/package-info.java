/**
 * 
 */
/**
 * @author DLEE
 * 
 */
@Module(name="module.json",moduleClass="Java")
package org.xmlsh.modules.java;
import org.xmlsh.annotations.Module;

