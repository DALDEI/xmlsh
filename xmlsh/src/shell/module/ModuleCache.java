package module;

import java.net.URI;

import org.xmlsh.util.NameValueMap;


/*
 * Global module type cache
 * Serves as factory class and resolver
 * 
 */
public class ModuleCache {

	// id -> ModuleClass 
	private NameValueMap<ModuleClass>  mModuleCache;
	
	 
	public ModuleCache() {
	  mModuleCache = new 	NameValueMap<>();
	}
	
	synchronized ModuleClass  findById( URI id ){
		
		return mModuleCache.get(id);
		
	}
	
	

}
