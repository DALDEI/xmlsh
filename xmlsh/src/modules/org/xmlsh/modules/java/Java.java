package org.xmlsh.modules.java;

import java.util.List;

import org.xmlsh.annotations.Function;
import org.xmlsh.core.AbstractBuiltinFunction;
import org.xmlsh.core.CoreException;
import org.xmlsh.core.XValue;
import org.xmlsh.sh.module.ModuleConfig;
import org.xmlsh.sh.module.PackageModule;
import org.xmlsh.sh.shell.Shell;
import org.xmlsh.util.JavaUtils;

@org.xmlsh.annotations.Module(name="java-types")
public class Java extends PackageModule {

	public Java(ModuleConfig config) throws CoreException {
		super(config);
		// TODO Auto-generated constructor stub
	}
	

	@Function( name = "getClass")
	public static class getClass extends AbstractBuiltinFunction
	{


	  @Override
	  public XValue run(Shell shell, List<XValue> args) throws Exception
	  {
		  requires(args.size() > 0 , "getClass( name ) ");
			return XValue.newXValue(JavaUtils.convertToClass(args.get(0), shell));
	  }
	}

}
