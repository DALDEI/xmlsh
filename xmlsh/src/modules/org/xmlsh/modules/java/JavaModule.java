package org.xmlsh.modules.java;

import java.util.List;

import org.xmlsh.annotations.Function;
import org.xmlsh.core.AbstractBuiltinFunction;
import org.xmlsh.core.CoreException;
import org.xmlsh.core.XClassLoader;
import org.xmlsh.core.XValue;
import org.xmlsh.sh.module.ModuleConfig;
import org.xmlsh.sh.module.PackageModule;
import org.xmlsh.sh.shell.Shell;
import org.xmlsh.types.TypeFamily;
import org.xmlsh.util.JavaUtils;

@org.xmlsh.annotations.Module(name="java")
public class JavaModule extends PackageModule {

	public JavaModule(ModuleConfig config, XClassLoader loader) throws CoreException {
		super(config, loader);
		// TODO Auto-generated constructor stub
	}
	

	@Function( name = "get-class" , names={"class"} )
	public static class getClass extends AbstractBuiltinFunction
	{

	  @Override
	  public XValue run(Shell shell, List<XValue> args) throws Exception
	  {
		  requires(args.size() > 0 , "( name ) ");
			return XValue.newXValue(JavaUtils.convertToClass(args.get(0), shell));
	  }
	}
	@Function( name="new" , names={"new", "jnew"} )
	public static class _new extends AbstractBuiltinFunction {

	    @Override
	    public XValue run(Shell shell, List<XValue> args) throws Exception {

	        Class<?> cls = JavaUtils.convertToClass(args.remove(0), shell);
	        XValue obj = null;
	        obj = JavaUtils.newXValue(cls, args);
	        return obj;


	    }

	}
}
