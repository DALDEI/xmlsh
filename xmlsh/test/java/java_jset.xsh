# Test of jset
import module java
# Simple string
. ../common
# empty string

jset -v s1 -c java.lang.String
echo -n loc() ""
xtype {$s1}
echo loc() String Value: $s1

# string with value
jset -v s1 -c java.lang.String "Hi"
echo loc() String Value: "(init)" $s1

# Append
jset -v s1 -o {$s1} -m concat " There"
echo loc() String Value: "(concat)"  $s1

# Replace (2 args)
jset -v s2 -o {$s1} -m replace "Hi" "Bye"
echo loc() String Value: "(replace)" $s2

# Integer values
jset -v i1 -o {$s1}-m length
echo -n loc() ""
xtype {$i1}
echo loc()  Length: $i1

# Import test
import java ../bin/xmlsh-test.jar

# Default constructor
jset -v t1 -c org.xmlsh.test.TestTypes
echo -n loc() ""
xtype {$t1}
jset -v m1 -o {$t1} -m getConstructor

# String constructor
jset -v t1 -c org.xmlsh.test.TestTypes "String"
jset -v m1 -o {$t1} -m getConstructor
echo loc() $m1

# Integer constructor
jset -v t1 -c org.xmlsh.test.TestTypes $i1
jset -v m1 -o {$t1} -m getConstructor
echo loc()  $m1


# Overloaded multi arg constructors
jset -v t2 -c org.xmlsh.test.TestTypes <[ 1 ]> "String"
jset -v m2 -o {$t2} -m getConstructor
echo loc() $m2





