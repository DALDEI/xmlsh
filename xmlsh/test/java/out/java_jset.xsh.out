[8] java.lang.String
[10] String Value:
[14] String Value: (init) Hi
[18] String Value: (concat) Hi There
[22] String Value: (replace) Bye There
[26] java.lang.Integer
[28] Length: 8
[35] org.xmlsh.test.TestTypes
[42] TestTypes(String)
[47] TestTypes(Integer)
[53] TestTypes(int,String)
